File: src/main/java/com/micabyte/android/app/LinearListView.java
package com.micabyte.android.app;

import android.annotation.TargetApi;
import android.content.Context;
import android.database.DataSetObserver;
import android.os.Build;
import android.util.AttributeSet;
import android.view.View;
import android.widget.Adapter;
import android.widget.LinearLayout;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

// android:orientation
public class LinearListView extends LinearLayout
{
    Adapter adapter;
    Observer observer = new Observer(this);

    public LinearListView(Context context)
    {
        super(context);
    }

    public LinearListView(Context context, AttributeSet attrs)
    {
        super(context, attrs);
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    public LinearListView(Context context, AttributeSet attrs, int defStyle)
    {
        super(context, attrs, defStyle);
    }

    public void setAdapter(Adapter adapter)
    {
        if (this.adapter != null)
            this.adapter.unregisterDataSetObserver(observer);

        this.adapter = adapter;
        adapter.registerDataSetObserver(observer);
        observer.onChanged();
    }

    private class Observer extends DataSetObserver
    {
        LinearListView context;

        public Observer(LinearListView context)
        {
            this.context = context;
        }

        @Override
        public void onChanged()
        {
            List<View> oldViews = new ArrayList<View>(context.getChildCount());

            for (int i = 0; i < context.getChildCount(); i++)
                oldViews.add(context.getChildAt(i));

            Iterator<View> iter = oldViews.iterator();

            context.removeAllViews();

            for (int i = 0; i < context.adapter.getCount(); i++)
            {
                View convertView = iter.hasNext() ? iter.next() : null;
                context.addView(context.adapter.getView(i, convertView, context));
            }
            super.onChanged();
        }

        @Override
        public void onInvalidated()
        {
            context.removeAllViews();
            super.onInvalidated();
        }
    }
}

File: src/main/java/com/micabytes/ui/component/LinearListView.kt
package com.micabytes.ui.component

import android.annotation.TargetApi
import android.content.Context
import android.database.DataSetObserver
import android.os.Build
import android.util.AttributeSet
import android.view.View
import android.widget.Adapter
import android.widget.BaseAdapter
import android.widget.LinearLayout
import android.widget.ListView
import androidx.databinding.BindingAdapter
import java.util.*

class LinearListView : LinearLayout {
  var adapter: Adapter? = null
    set(adp) {
      if (this.adapter != null) {
        this.adapter!!.unregisterDataSetObserver(observer)
      }
      field = adp
      adp?.registerDataSetObserver(observer)
      observer.onChanged()
    }
  private val observer = Observer(this)

  constructor(context: Context) : super(context)

  constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

  @TargetApi(Build.VERSION_CODES.HONEYCOMB)
  constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)

  private class Observer internal constructor(internal val listView: LinearListView) : DataSetObserver() {

    override fun onChanged() {
      val oldViews = ArrayList<View>(listView.childCount)
      (0 until listView.childCount).mapTo(oldViews) { listView.getChildAt(it) }
      val itr = oldViews.iterator()
      listView.removeAllViews()
      for (i in 0 until listView.adapter!!.count) {
        val convertView = if (itr.hasNext()) itr.next() else null
        listView.addView(listView.adapter!!.getView(i, convertView, listView))
      }
      super.onChanged()
    }

    override fun onInvalidated() {
      listView.removeAllViews()
      super.onInvalidated()
    }
  }

  companion object {

    @JvmStatic
    @BindingAdapter("adapter")
    fun bindList(view: LinearListView, adapter: BaseAdapter) {
      view.adapter = adapter
    }

    @JvmStatic
    @BindingAdapter("adapter")
    fun bindList(view: ListView, adapter: BaseAdapter) {
      view.adapter = adapter
    }

  }

}
